src/mame/drivers/kinst.cpp

```diff
@@ -374,12 +382,31 @@ INTERRUPT_GEN_MEMBER(kinst_state::irq0_start)
 
 uint32_t kinst_state::ide_r(offs_t offset, uint32_t mem_mask)
 {
-	return m_ata->cs0_r(offset / 2, mem_mask);
+  if ((offset==0x0c)&&(prot_sel)) /* read register from kinst2uk protection CPLD */
+    return (prot_sel<<4)|(prot_cnt^prot_rega^prot_regb); /* return last upper bits written with xor of counter and registers in lower 4 bits */
+  else
+    return m_ata->cs0_r(offset / 2, mem_mask);
 }
 
 
 void kinst_state::ide_w(offs_t offset, uint32_t data, uint32_t mem_mask)
 {
+  if ((offset==0x0c)&&(data&0x10)) /* write to IDE slave (kinst2uk protection CPLD) */
+  {
+    if (data&0x40) /* D6 high */
+    {
+      prot_sel=0x05; /* select IDE slave (D6 high) */
+      prot_rega=data&0x0F; /* store lower 4 bits in register a */
+    }
+    else /* D6 low */
+    {
+      prot_sel=0x01; /* select IDE slave (D6 low) */
+      prot_regb=data&0x0F; /* store lower 4 bits in register b */
+      prot_cnt=(prot_cnt+1)&0x0F; /* every write here increments counter */
+    }
+  }
+  else if ((offset==0x0c)&&(!(data&0x10))) /* write to IDE master */
+    prot_sel=0; /* deselect IDE slave */
 	m_ata->cs0_w(offset / 2, data, mem_mask);
```

Comment:
Please follow out coding conventions.  Use one tab indent per level of scope, use lowercase hex digits consistently, and don’t code like your spacebar is broken, and don’t code like your space bar is broken.

Also, the CPLD is only present for the upgrade kit, and hence shouldn’t be emulated all the time.  It also affects only a single offset.

* Make separate read/write functions for handing the offset affected by the CPLD.  You can call through to the regular `ide_r` and `ide_w` from your read/write functions when the CPLD isn’t intercepting the access.
* Make a derived address map function for the upgrade kit, and install your handlers over the top at the offset they affect.
* Make a derived machine configuration function for the upgrade kit and set your modified address map.
* Use this machine configuration function in the `GAME` macro for the upgrade kit.